### ISSUE TITLE
Some chunks not saved after long use.

### ISSUE THREAD

**Issue by Esper89** [mircokroon/minecraft-world-downloader#90]
After using it for a while on one server, the chunks seem to have stopped saving, despite the GUI showing them as saved. After stopping the world downloader and opening it again, it shows many chunks are no longer saved, and upon loading the same world in WorldPainter it shows the same thing. Possibly related to #89 since those crashes happened at the same time as this. Possibly related to the fact that the world is 92.3 MB and it was mostly downloaded all at once without restarting the downloader.

Allegedly saved chunks before restarting the world downloader:
![Previously saved chunks](https://user-images.githubusercontent.com/69338029/102020421-26ed7280-3d47-11eb-9950-0fbff44f32ae.png)
WorldPainter loading the world afterwards:
![Saved chunks after restarting](https://user-images.githubusercontent.com/69338029/102020424-281e9f80-3d47-11eb-96e9-d4f78a229ba3.png)

[Here](https://envl.app/esper8989/NwhOIz!caVIDO5GmGHRjC6B)'s the world file. It's too big to be uploaded to GitHub.

The chunks may have been saved and just not recognized as valid chunks by both the world downloader and WorldPainter. Opening the world normally causes Minecraft to crash, but that's an issue even when only a few chunks are saved - it's something to do with an invalid banner entity, and probably has to do with the Wynncraft server having mods that aren't present on the regular singleplayer server.

There were also several other (smaller) areas, thousands of blocks away from the parts shown in the screenshot above. Those areas also looked like they were saved but weren't.

**Comment by mircokroon** [comment]
Were there any messages in the World Downloader terminal? 

Also, did you touch the world in any way during the download? This could happen for sure if you for example opened the world in Minecraft while the downloader was writing to it

**Comment by Esper89** [comment]
I don't think I saw anything in the terminal output; it's possible that I missed something but I know that I read it all and I would have remembered something strange. There were many, many messages about clients and servers "probably disconnecting" but that was because I was actually disconnecting and connecting again (occasionally because of the aforementioned crashes).

I have yet to open the world with anything but the downloader, so I definitely didn't touch the world while it was downloading. I copied it to my saves directory and opened the copy with Minecraft, and I did that while the downloader wasn't running anyway.

And yeah, a GUI for something like that probably would have saved me half an hour of running along the edges of mountains with "show chunk borders" enabled while it wasn't even saving the chunks. Oh well, now I know that I should restart the program more often to check and make sure.

Something I neglected to mention was that I closed the program just by sending `^C` to it. I figured that's the appropriate way to close it (since there are no other obvious ways to do so) but if that could have been the cause of this issue then I'd like to know how else I'm supposed to close it.

**Comment by mircokroon** [comment]
Apologies for the wasted time!

Technically the correct way to close the application is to right-click in the GUI and clicking on "Save & exit". This would probably have saved your chunks if they had still been in memory, but I imagine you may have run out of memory if that had been the case for so many chunks.
Closing the application this way can cause _some_ chunks to get lost, as it only saves every ~20 seconds usually. It definitely would not explain half an hour of them vanishing though. 

I don't really see a reason for long usage to cause these problems, so I think for now the best way to go is to add something that indicates in the GUI which chunks are saved so at least the user will be able to notice if something is wrong. I'll have a look at your world file soon too, to see if that reveals something.

**Comment by Esper89** [comment]
Thank you, I hadn't thought to right-click the GUI. That probably never would have occurred to me, since closing the GUI doesn't close the program, I figured it was mostly independent.

If it has to do with running out of memory, would increasing the memory limit in Java help with this issue? I'm thinking something like `-Xmx4096m`, but I don't know very much about how that works.

**Comment by mircokroon** [comment]
If that had been the problem I think it would have fully crashed, so probably it's not that. 

The main reason I mentioned that was that chunks are usually only unloaded from the downloader's memory once they have been written to the disk. Looking at your world, I don't see the chunks that were supposed to have been downloaded, so if they weren't written they must have been in memory, but it seems like way too much data to have been in memory all at once. 

Basically I have no idea. In any case I'll add the saved-or-not indicator so at least such problems will be clear in the future.

**Comment by Esper89** [comment]
Okay, that makes sense. The saved amount was only around 92.3 MB anyway, so I don't see how the rest could have gone over the maximum. Thanks for your help!

**Comment by mircokroon** [comment]
I added least added the marking of unsaved chunks in the latest version:
![image](https://user-images.githubusercontent.com/23699979/102023783-2926ef80-3d8e-11eb-8444-f346c548385a.png)

It's unfortunately I can't find a fix for the core problem, but maybe some sort of reproducible scenario will show up eventually now that we have a way to see it happening.

**Comment by Esper89** [comment]
If I notice it happening again I'll try and see more information. On that note, is there any way to see more verbose logs then what it normally sends to stdout and stderr, to get more information?

**Comment by mircokroon** [comment]
Not really, it's supposed to print quite clearly at the very least when something goes wrong though. I hope that whatever could cause so many chunks to go unsaved would fall under that.

Verbose logging probably should be added at some point.

**Comment by mircokroon** [comment]
Seeing as this has not been update in a while, I'll close it and hope that it's been resolved somehow. If it happens again feel free to re-open or something.
